### PR TITLE
First prototype of RAG UI.

### PR DESCRIPTION
**Pull request by nsthorat**
Demo: https://lilacai-nikhil-staging.hf.space/rag#&datasetNamespace=%22local%22&datasetName=%22mosaic-chat-v2%22&path=%5B%22response%22%5D&embedding=%22gte-small%22&query=%22How%20tall%20is%20the%20empire%20state%20building%3F%22

image

### CODE REVIEW

File: lilac/router_rag.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                      'Please install it with `pip install openai`.')
+
+  openai.api_key = env('OPENAI_API_KEY')
+  completion = openai.ChatCompletion.create(
+    model='gpt-3.5-turbo-0613',
+    functions=[Answer.openai_schema],
+    messages=[
+      {
+        'role': 'system',
+        'content': 'You must call the `Answer` function with the generated examples',
```
Comment (dsmilkov) [inline lilac/router_rag.py:34]: we should repharse this. You must call the `Answer` function with the answer to the question?

**Review by dsmilkov — APPROVED** [review]
Amazing work!!

File: lilac/router_rag.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+router = APIRouter(route_class=RouteErrorHandler)
+
+
+class Answer(OpenAISchema):
+  """Generated text examples."""
+  answer: str = Field(..., description='The answer to the question, given the context and query.')
+
+
+@router.get('/generate_answer')
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply lilac/router_rag.py:19]: Done

File: lilac/router_rag.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+def generate_answer(prompt: str) -> str:
+  """Generate positive examples for a given concept using an LLM model."""
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply lilac/router_rag.py:21]: Done

File: lilac/router_rag.py (hunk 1 above)
Comment (nsthorat) [reply lilac/router_rag.py]: Done, changed to completion

File: web/blueprint/src/app.css [hunk 4]
```diff
@@ -113,8 +113,14 @@ button:disabled {
   @apply align-middle;
 }
 
+/** Rag components */
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply web/blueprint/src/app.css:116]: This is CSS that is shared between multiple components, and I didn't think it made sense to just pull the header out to a separate component

File: web/blueprint/src/lib/stores/ragViewStore.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  datasetNamespace: string | null;
+  datasetName: string | null;
+  path: Path | null;
+  embedding: string | null;
+
+  query: string | null;
+  promptTemplate: string;
+  topK: number;
+  semanticSimilarityThreshold: number;
+  windowSizeChunks: number;
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply web/blueprint/src/lib/stores/ragViewStore.ts:21]: Done

File: web/blueprint/src/lib/stores/ragViewStore.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    path: null,
+    embedding: null,
+    query: null,
+    promptTemplate: `Context information is below.
+---------------------
+{context_str}
+---------------------
+Given the context information and not prior knowledge, answer the query.
+Query: {query_str}
+Answer: \
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply web/blueprint/src/lib/stores/ragViewStore.ts:38]: Done

File: web/blueprint/src/routes/rag/+page.svelte [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            ragViewStore.setDatasetPathEmbedding(dataset, path, embedding);
+          }
+        }}
+      />
+    {/if}
+  </div>
+
+  <div class="mx-4 mb-8 mt-2 flex h-full w-full flex-col gap-y-24 px-4">
+    <div class="mt-6 flex w-full flex-row gap-x-10">
+      <!-- Input question -->
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply web/blueprint/src/routes/rag/+page.svelte:130]: Done, centering looked weird so there's just a hole in the top-right. probably fine for now.

File: lilac/router_rag.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  try:
+    import openai
+  except ImportError:
+    raise ImportError('Could not import the "openai" python package. '
+                      'Please install it with `pip install openai`.')
+
+  openai.api_key = env('OPENAI_API_KEY')
```
[the remark this answers is not in the document]
Comment (nsthorat) [reply lilac/router_rag.py:28]: Done

File: lilac/router_rag.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dsmilkov) [reply lilac/router_rag.py:21]: looks like pydoc is still the old one

**Pull request merged**